### PR TITLE
Add auth0 with scoped examples

### PR DESCRIPTION
**Pull request by EdenGottlieb**
This PR adds auth0 with basic examples on how to use it.

For admin & creation of new oauth scopes, you can contact me and I'll add you to Auth0

### CODE REVIEW

**Review by YanDav78 — CHANGES_REQUESTED** [review]
I'm not familiar with this library, please add comments so we can understand better.
Thanks

File: app/services/auth.py [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+    def __init__(self, detail: str, **kwargs):
+        """Returns HTTP 403"""
+        super().__init__(status.HTTP_403_FORBIDDEN, detail=detail)
+
+
+class UnauthenticatedException(HTTPException):
+    def __init__(self):
+        super().__init__(
+            status_code=status.HTTP_401_UNAUTHORIZED, detail="Requires authentication"
+        )
```
Comment (itamar-cord) [inline app/services/auth.py:42]: are we sure this is the correct error mapping? shouldn't UnauthorizedException return an unaurthorized status?

File: app/services/auth.py [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+
+class VerifyToken:
+    """Does all the token verification using PyJWT"""
+
+    def __init__(self):
+        self.config = get_auth0_config()
+
+        # This gets the JWKS from a given URL and does processing so you can
+        # use any of the keys available
+        jwks_url = f'https://{self.config["DOMAIN"]}/.well-known/jwks.json'
```
Comment (itamar-cord) [inline app/services/auth.py:51]: can you log somewhere which env variables we will need to configure in the prod environment when it is up so we wont need to debug it?

File: app/services/auth.py [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+        self.jwks_client = jwt.PyJWKClient(jwks_url)
+
+    async def verify(self,
+                     security_scopes: SecurityScopes,
+                     token: Optional[HTTPAuthorizationCredentials] = Depends(HTTPBearer())
+                     ):
+        if token is None:
+            raise UnauthenticatedException
```
Comment (itamar-cord) [inline app/services/auth.py:61]: do we want to add a message about the missing token to help solve this error?

File: requirements.txt [hunk 4]
```diff
@@ -1,5 +1,6 @@
 fastapi
 uvicorn
+PyJWT
```
Comment (itamar-cord) [inline requirements.txt:3]: do we want to put a specific version?

File: app/services/auth.py (hunk 1 above)
Comment (EdenGottlieb) [reply app/services/auth.py:42]: WDYM by unauthorized status?
403 is for when you're logged in but not allowed to do something which afaik is the same as unauthorized
I took this code from auth0's example

File: app/services/auth.py (hunk 2 above)
Comment (EdenGottlieb) [reply app/services/auth.py:51]: It's in .config which is committed to the repo
Obviously not the best practice but I believe it's OK for now

File: app/services/auth.py (hunk 3 above)
Comment (EdenGottlieb) [reply app/services/auth.py:61]: 401 indicates the user is missing authentication (or missing a token)

File: requirements.txt (hunk 4 above)
Comment (EdenGottlieb) [reply requirements.txt:3]: All of the other deps don't have versions so I followed that norm.

File: app/services/auth.py (hunk 1 above)
Comment (itamar-cord) [reply app/services/auth.py:42]: I mean, 
UnauthorizedException-> HTTP_403_FORBIDDEN
UnauthenticatedException -> HTTP_401_UNAUTHORIZED

look unintuitive but don't care really

File: app/services/auth.py (hunk 3 above)
Comment (itamar-cord) [reply app/services/auth.py:61]: can also mean that the token is invalid right? is there a difference in the code?

Comment (EdenGottlieb) [reply app/services/auth.py:61]: Invalid token gives you 403 (lines 69 & 71 & 85)

**Review by itamar-cord — APPROVED** [review]

**Comment by SereneSound** [comment]
@itamar-cord  @YanDav78 Why is this not merged yet?

**Comment by EdenGottlieb** [comment]
> @itamar-cord  @YanDav78 Why is this not merged yet?

Apparently our convention is that the poor request creator needs to merge it  :)

**Pull request merged**